library/old_protocol/old_protocol.cpp

```diff
@@ -0,0 +1,170 @@
+	switch (cmd.cmd) {
+	case DNET_CMD_LOOKUP_NEW:
+		return translate_lookup_request(st, cmd);
+	default:
+		// Must never reach this code, due to is_supported_message() filter called before
+		return -ENOTSUP;
+	}
+}
+
+int old_protocol::recv_response(dnet_net_state *st, const dnet_cmd &cmd, data_pointer &&body) {
+	int err = 0;
+	auto it = repliers_table_.find(cmd.trans);
+	auto &repliers = it->second;
+
+	if (cmd.status) {
+		err = repliers.on_reply_error(cmd.status);
+
+	} else {
+		switch (cmd.cmd) {
+		case DNET_CMD_LOOKUP_NEW:
+			{
+				std::unique_ptr<n2_message> msg;
+				err = deserialize_lookup_response(msg, st, cmd, std::move(body));
+				if (err)
+					break;
+
+				err = repliers.on_reply(std::move(msg));
+			}
+			break;
+		default:
+			// Must never reach this code, due to is_supported_message() filter called before
+			err = -ENOTSUP;
+		}
+	}
+
+	repliers_table_.erase(it);
+	return err;
+}
+
+int old_protocol::schedule_request_info(dnet_net_state *st,
+                                        std::unique_ptr<n2_request_info> &&request_info) {
+	request_info->cmd = request_info->request->cmd;
+	return on_request(st, std::move(request_info));
+}
+
+int old_protocol::translate_lookup_request(dnet_net_state *st, const dnet_cmd &cmd) {
+	std::unique_ptr<n2_request_info> request_info(new n2_request_info);
+
+	int err = deserialize_lookup_request(request_info->request, st, cmd);
+	if (err)
+		return err;
+
+	auto replier = std::make_shared<lookup_replier>(st, request_info->request->cmd);
+	request_info->repliers.on_reply = std::bind(&lookup_replier::reply, replier, std::placeholders::_1);
+	request_info->repliers.on_reply_error = std::bind(&lookup_replier::reply_error, replier, std::placeholders::_1);
+
+	return schedule_request_info(st, std::move(request_info));
+}
+
+}}} // namespace ioremap::elliptics::n2
+
+extern "C" {
+
+int n2_old_protocol_io_start(struct dnet_node *n) {
+	auto impl = [io = n->io] {
+		io->old_protocol = new n2_old_protocol_io;
+		return 0;
+	};
+	return c_exception_guard(impl, n, __FUNCTION__);
+}
+
+void n2_old_protocol_io_stop(struct dnet_node *n) {
+	auto impl = [io = n->io] {
+		delete io->old_protocol;
+		io->old_protocol = nullptr;
+		return 0;
+	};
+	c_exception_guard(impl, n, __FUNCTION__);
+}
+
+int n2_old_protocol_rcvbuf_create(struct dnet_net_state *st) {
+	st->rcv_buffer = new n2_recv_buffer;
+	return 0;
+}
+
+void n2_old_protocol_rcvbuf_destroy(struct dnet_net_state *st) {
+	delete st->rcv_buffer;
+	st->rcv_buffer = nullptr;
+}
+
+bool n2_old_protocol_is_supported_message(struct dnet_net_state *st) {
+	const dnet_cmd *cmd = &st->rcv_cmd;
+
+	// Replies addressed to client are currently passed via old mechanic. This condition branch will be removed
+	// after client supports new mechanic for DNET_CMD_LOOKUP_NEW command.
+	if (cmd->flags & DNET_FLAGS_REPLY) {
+		std::unique_ptr<dnet_trans, void (*)(dnet_trans *)>
+		        t(dnet_trans_search(st, cmd->trans), &dnet_trans_put);
```

Comment:
`dnet_trans_search` should be called with locked `st->trans_lock`.

Reply:
Fixed.